torch/csrc/Module.cpp

```diff
@@ -570,6 +573,118 @@ PyObject *THPModule_fromDLPack(PyObject *_unused, PyObject *data)
+{                                                                             \
+    write(fileno(stderr), ERROR_MSG, strlen(ERROR_MSG));                      \
+    _exit(EXIT_FAILURE);                                                      \
+}
+
+SIGNAL_HANDLER(SIGBUS, handler_SIGBUS, "ERROR: Unexpected bus error encountered in worker. "
+  "This might be caused by insufficient shared memory (shm).\n");
+SIGNAL_HANDLER(SIGSEGV, handler_SIGSEGV, "ERROR: Unexpected segmentation fault encountered in worker.\n");
+
+static std::mutex worker_pid_mutex;
+static std::vector<pid_t> worker_pid_vec = {};
+// The following are needed since std::vector is not asynchronous safe.
+static pid_t *worker_pids;
+static size_t num_worker_pids = 0;
+
+PyObject *THPModule_setWorkerSignalHandlers(PyObject *module, PyObject *arg) {
+    signal(SIGBUS, handler_SIGBUS);
+    signal(SIGSEGV, handler_SIGSEGV);
+    Py_RETURN_NONE;
+}
+
+// SIGCHLD hander should be registered on main loader process to catch any
+// worker failing. SIGALRM handler is needed for implementing timeout.
+// Python handles are _set_main_signal_handers() and
+// _remove_main_signal_handers().
+static void handler_SIGCHLD_main(int sig) {
+  int status;
+  pid_t p;
+  pid_t *pid_ptr;
+
+  // The flags and status checks ensure that we are really observing a child
+  // exiting, rather than other cases such as SIGSTOP and SIGCONT.
+  // https://stackoverflow.com/a/40707100
+  while ((p = waitpid(-1, &status, WNOHANG|WUNTRACED|WCONTINUED)) > 0) {
+    if (WIFCONTINUED(status) || WIFSTOPPED(status))
+      continue;
+    if (WIFEXITED(status) != 0 && WEXITSTATUS(status) == 0)
+      continue;
+    // child must have exited with signal/error, check if it is one of the pid
+    // we care about.
+    pid_ptr = worker_pids;
+    for (size_t i = 0; i < num_worker_pids; i++) {
+      if (*pid_ptr == p)
+        _exit(EXIT_FAILURE);
+      pid_ptr++;
+    }
+  }
+}
+
+SIGNAL_HANDLER(SIGALRM, handler_SIGALRM, "ERROR: Time out when fetching data in DataLoader.\n")
+
+// We don't want to exit on any SIGCHLD from any child. child_pids is a sequence
+// of pids we are interested in.
+PyObject *THPModule_setMainSignalHandlers(PyObject *module, PyObject *child_pids) {
+  auto tuple = PyTuple_Check(child_pids);
+  THPUtils_assert(tuple || PyList_Check(child_pids), "_set_main_signal_handler "
+        "expects a tuple or list, but got %s", THPUtils_typename(child_pids));
+
+  auto size = tuple ? PyTuple_GET_SIZE(child_pids) : PyList_GET_SIZE(child_pids);
+  {
+    std::unique_lock<std::mutex> lock(worker_pid_mutex);
+    for (int idx = 0; idx < size; idx++) {
+      PyObject* obj = tuple ? PyTuple_GET_ITEM(child_pids, idx) : PyList_GET_ITEM(child_pids, idx);
+      worker_pid_vec.push_back((pid_t) THPUtils_unpackLong(obj));
+    }
+    worker_pids = &worker_pid_vec[0];
+    num_worker_pids = worker_pid_vec.size();
+  }
+
+  signal(SIGCHLD, handler_SIGCHLD_main);
```

Comment:
Does this ensure that the signal is only handled by the main thread?

Reply:
Good question. Python's `signal` module handler always ensures main thread gets the signal. Unfortunately, on c++ in POSIX we don't have such guarantee. I can add some signal blocking to ensure that threads spawned after this won' t handle the signal, but I'm not sure what I we can do with threads spawned before this. I'll investigate more tomorrow.

Reply:
Does it really matter which thread handles the signal? `_exit` should abort the whole program, right?

Reply:
@apaszke is right. Somehow I mistakenly thought each thread has its separate handler.